### PR TITLE
Ensure that mean and standard deviation are broadcastable in standardization preprocessors

### LINKED ISSUE #268: PyTorchClassifier preprocessing on Imagenet Images
**Describe the bug**
Image preprocessing with the PyTorchClassifier on imagenet images does not seem to broadcast the means and stds correctly. Or maybe I'm just missing something?

**To Reproduce**
```
import torch
from torchvision import models

from art.classifiers import PyTorchClassifier
from art.attacks import ProjectedGradientDescent

# load and preprocess imagenet images -- values in [0,1]
images = load_preprocess_images(im_paths)

alexnet = models.alexnet(pretrained=True)

loss = torch.nn.CrossEntropyLoss()
optimizer = torch.optim.SGD(model.parameters(), .1, momentum=0.9, weight_decay=1e-4)
wrapped_model = PyTorchClassifier(
    model=alexnet, 
    loss=loss, 
    optimizer=optimizer, 
    clip_values=(0, 1),
    input_shape=(3,224,224),
    nb_classes=1000,
    channel_index=1,
    preprocessing=((0.485, 0.456, 0.406), (0.229, 0.224, 0.225))
)

eps = 5
norm = 2
steps = 10
attack = ProjectedGradientDescent(wrapped_model, norm=norm, eps_step=eps/(.2*steps), eps=eps, max_iter=steps)

adv_images = attack.generate(x = images)
```
returns the error: 
```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-65-c7ccecb30513> in <module>
     28 attack = ProjectedGradientDescent(wrapped_model, norm=norm, eps_step=eps/(.2*steps), eps=eps, max_iter=steps)
     29 
---> 30 adv_images = attack.generate(x = images)

~/anaconda3/envs/brainscore2/lib/python3.6/site-packages/art/attacks/projected_gradient_descent.py in generate(self, x, y, **kwargs)
    108 
    109             # Use model predictions as correct outputs
--> 110             targets = get_labels_np_array(self.classifier.predict(x, batch_size=self.batch_size))
    111         else:
    112             targets = y

~/anaconda3/envs/brainscore2/lib/python3.6/site-packages/art/classifiers/pytorch.py in predict(self, x, batch_size, **kwargs)
    104 
    105         # Apply preprocessing
--> 106         x_preprocessed, _ = self._apply_preprocessing(x, y=None, fit=False)
    107 
    108         # Run prediction with batch processing

~/anaconda3/envs/brainscore2/lib/python3.6/site-packages/art/classifiers/classifier.py in _apply_preprocessing(self, x, y, fit)
    166         y = check_and_transform_label_format(y, self.nb_classes())
    167         x_preprocessed, y_preprocessed = self._apply_preprocessing_defences(x, y, fit=fit)
--> 168         x_preprocessed = self._apply_preprocessing_standardisation(x_preprocessed)
    169         return x_preprocessed, y_preprocessed
    170 

~/anaconda3/envs/brainscore2/lib/python3.6/site-packages/art/classifiers/classifier.py in _apply_preprocessing_standardisation(self, x)
    208             div = np.asarray(div, dtype=x.dtype)
    209 
--> 210             res = x - sub
    211             res = res / div
    212 

ValueError: operands could not be broadcast together with shapes (3,3,224,224) (3,)
```

**Expected behavior**
The images should have the channel means subtracted off, and then be divide by the channel standard deviations -- given the input shape, this would require the means and stds to be broadcasted correctly. a brief inspection of the code suggests the broadcasting is not being handled correctly, but maybe I'm missing something?

**System information (please complete the following information):**
-    Fedora Linux
-    Python 3.6.7
-    ART 1.0.1
-    PyTorch 1.2.0

### PR DESCRIPTION
**Pull request by hesseltuinhof**
# Description

This PR adds functionality that aims to catch situations where the mean and standard deviation are not broadcastable w.r.t. to the input batch. It then reshapes the mean and standard deviation such that addition and division operations in the standardization preprocessors work properly.

Fixes #268, #569

## Type of change

Please check all relevant options.

- [x] Improvement (non-breaking)

# Testing

**Test Configuration**:
- Fedora 33
- Python 3.8.7
- ART 1.6.0 development

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: art/preprocessing/standardisation_mean_std/utils.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
+# MIT License
+#
+# Copyright (C) The Adversarial Robustness Toolbox (ART) Authors 2020
```
Comment (beat-buesser) [inline art/preprocessing/standardisation_mean_std/utils.py:3]:
```suggestion
# Copyright (C) The Adversarial Robustness Toolbox (ART) Authors 2021
```

File: art/preprocessing/standardisation_mean_std/standardisation_mean_std_tensorflow.py [hunk 2]
```diff
@@ -40,7 +41,11 @@ class StandardisationMeanStdTensorFlowV2(PreprocessorTensorFlowV2):
     params = ["mean", "std", "apply_fit", "apply_predict"]
 
     def __init__(
-        self, mean: float = 0.0, std: float = 1.0, apply_fit: bool = True, apply_predict: bool = True,
+        self,
+        mean: Union[float, Sequence, np.ndarray] = 0.0,
+        std: Union[float, Sequence, np.ndarray] = 1.0,
```
Comment (beat-buesser) [inline art/preprocessing/standardisation_mean_std/standardisation_mean_std_tensorflow.py:46]: I agree with adding `np.ndarray` but is `Sequence` required?

File: art/preprocessing/standardisation_mean_std/standardisation_mean_std_tensorflow.py [hunk 3]
```diff
@@ -49,19 +54,21 @@ def __init__(
...
+        self.std = np.asarray(std, dtype=ART_NUMPY_DTYPE)
         self._check_params()
 
     def forward(self, x: "tf.Tensor", y: Optional["tf.Tensor"] = None) -> Tuple["tf.Tensor", Optional["tf.Tensor"]]:
         """
         Apply standardisation with mean and standard deviation to input `x`.
+
+        :param x: Input samples to standardise of shapes `NCHW`, `NHWC`, `NCFHW` or `NFHWC`.
+        :param y: Label data, will not be affected by this preprocessing.
+        :return: Standardise input samples and unmodified labels.
```
Comment (beat-buesser) [inline art/preprocessing/standardisation_mean_std/standardisation_mean_std_tensorflow.py:67]:
```suggestion
        :return: Standardised input samples and unmodified labels.
```

File: art/preprocessing/standardisation_mean_std/standardisation_mean_std_tensorflow.py [hunk 4]
```diff
@@ -49,19 +54,21 @@ def __init__(
...
         """
         import tensorflow as tf  # lgtm [py/repeated-import]
 
-        mean = np.asarray(self.mean, dtype=ART_NUMPY_DTYPE)
-        std = np.asarray(self.std, dtype=ART_NUMPY_DTYPE)
-
+        mean, std = broadcastable_mean_std(x, self.mean, self.std)
```
Comment (beat-buesser) [inline art/preprocessing/standardisation_mean_std/standardisation_mean_std_tensorflow.py:71]: Would it be possible to run `broadcastable_mean_std` outside of method `forward`? It should only be required once.

File: art/preprocessing/standardisation_mean_std/standardisation_mean_std_tensorflow.py [hunk 5]
```diff
@@ -49,19 +54,21 @@ def __init__(
...
-        self.std = std
+        self.mean = np.asarray(mean, dtype=ART_NUMPY_DTYPE)
+        self.std = np.asarray(std, dtype=ART_NUMPY_DTYPE)
         self._check_params()
 
     def forward(self, x: "tf.Tensor", y: Optional["tf.Tensor"] = None) -> Tuple["tf.Tensor", Optional["tf.Tensor"]]:
         """
         Apply standardisation with mean and standard deviation to input `x`.
+
+        :param x: Input samples to standardise of shapes `NCHW`, `NHWC`, `NCFHW` or `NFHWC`.
```
Comment (beat-buesser) [inline art/preprocessing/standardisation_mean_std/standardisation_mean_std_tensorflow.py:65]: Does this new code also work with tabular data?

File: art/preprocessing/standardisation_mean_std/standardisation_mean_std_pytorch.py [hunk 6]
```diff
@@ -69,14 +73,16 @@ def forward(
     ) -> Tuple["torch.Tensor", Optional["torch.Tensor"]]:
         """
         Apply standardisation with mean and standard deviation to input `x`.
+
+        :param x: Input data for which the gradient is estimated. First dimension is the batch size.
+        :param grad: Gradient value so far.
+        :return: The gradient (estimate) of the defence.
```
Comment (beat-buesser) [inline art/preprocessing/standardisation_mean_std/standardisation_mean_std_pytorch.py:79]: I think part of this docsting needs updates, the version for tensorflow looks good.

File: art/preprocessing/standardisation_mean_std/standardisation_mean_std_pytorch.py [hunk 7]
```diff
@@ -69,14 +73,16 @@ def forward(
...
         """
         import torch  # lgtm [py/repeated-import]
 
-        mean = torch.tensor(self.mean, device=self._device)
-        std = torch.tensor(self.std, device=self._device)
-
-        x_norm = x - mean
-        x_norm = x_norm / std
+        mean, std = broadcastable_mean_std(x, self.mean, self.std)
```
Comment (beat-buesser) [inline art/preprocessing/standardisation_mean_std/standardisation_mean_std_pytorch.py:83]: Would it be possible to run broadcastable_mean_std outside of method forward? It should only be required once.

File: art/preprocessing/standardisation_mean_std/standardisation_mean_std_pytorch.py [hunk 8]
```diff
@@ -38,8 +42,8 @@ class StandardisationMeanStdPyTorch(PreprocessorPyTorch):
 
     def __init__(
         self,
-        mean: float = 0.0,
-        std: float = 1.0,
+        mean: Union[float, Sequence, np.ndarray] = 0.0,
+        std: Union[float, Sequence, np.ndarray] = 1.0,
```
Comment (beat-buesser) [inline art/preprocessing/standardisation_mean_std/standardisation_mean_std_pytorch.py:46]: I agree with adding np.ndarray but is Sequence required?

File: art/preprocessing/standardisation_mean_std/standardisation_mean_std.py [hunk 9]
```diff
@@ -38,7 +38,11 @@ class StandardisationMeanStd(Preprocessor):
     params = ["mean", "std", "apply_fit", "apply_predict"]
 
     def __init__(
-        self, mean: float = 0.0, std: float = 1.0, apply_fit: bool = True, apply_predict: bool = True,
+        self,
+        mean: Union[float, Sequence, np.ndarray] = 0.0,
+        std: Union[float, Sequence, np.ndarray] = 1.0,
```
Comment (beat-buesser) [inline art/preprocessing/standardisation_mean_std/standardisation_mean_std.py:43]: I agree with adding np.ndarray but is Sequence required?

File: art/preprocessing/standardisation_mean_std/standardisation_mean_std.py [hunk 10]
```diff
@@ -66,18 +70,24 @@ def __call__(self, x: np.ndarray, y: Optional[np.ndarray] = None,) -> Tuple[np.n
                 "np.float32.".format(x.dtype)
             )
 
-        mean = np.asarray(self.mean, dtype=ART_NUMPY_DTYPE)
-        std = np.asarray(self.std, dtype=ART_NUMPY_DTYPE)
-
+        mean, std = broadcastable_mean_std(x, self.mean, self.std)
```
Comment (beat-buesser) [inline art/preprocessing/standardisation_mean_std/standardisation_mean_std.py:73]: Would it be possible to run `broadcastable_mean_std` outside of method forward? It should only be required once.

File: art/preprocessing/standardisation_mean_std/standardisation_mean_std.py [hunk 11]
```diff
@@ -66,18 +70,24 @@ def __call__(self, x: np.ndarray, y: Optional[np.ndarray] = None,) -> Tuple[np.n
...
+        Provide an estimate of the gradients of preprocessor for the backward pass. If the preprocessor is not
+        differentiable, this is an estimate of the gradient, most often replacing the computation performed by the
+        preprocessor with the identity function (the default).
 
-        std = np.asarray(self.std, dtype=ART_NUMPY_DTYPE)
+        :param x: Input data for which the gradient is estimated. First dimension is the batch size.
+        :param grad: Gradient value so far.
+        :return: The gradient (estimate) of the defence.
+        """
+        _, std = broadcastable_mean_std(x, self.mean, self.std)
```
Comment (beat-buesser) [inline art/preprocessing/standardisation_mean_std/standardisation_mean_std.py:90]: Would it be possible to run `broadcastable_mean_std` outside of method `estimate_gradient`? It should only be required once.

**Review by beat-buesser — CHANGES_REQUESTED** [review]
Hi @hesseltuinhof Thank you very much for your pull request solving a long standing issue. I have a few questions and comments, some are duplicates and you only need to respond to the first of each. What do you think?

File: art/preprocessing/standardisation_mean_std/standardisation_mean_std_tensorflow.py (hunk 2 above)
Comment (hesseltuinhof) [reply art/preprocessing/standardisation_mean_std/standardisation_mean_std_tensorflow.py:46]: I removed `Sequence` type.

File: art/preprocessing/standardisation_mean_std/standardisation_mean_std_tensorflow.py (hunk 4 above)
Comment (hesseltuinhof) [reply art/preprocessing/standardisation_mean_std/standardisation_mean_std_tensorflow.py:71]: I don't see how we could move the call outside `forward`, e.g. into the `__init__` method. `broadcastable_mean_std` requires knowledge of input data `x` in order to infer how to reshape `mean` and `std`.

File: art/preprocessing/standardisation_mean_std/standardisation_mean_std_tensorflow.py (hunk 5 above)
Comment (hesseltuinhof) [reply art/preprocessing/standardisation_mean_std/standardisation_mean_std_tensorflow.py:65]: This docstring addition is legacy as I copied it from its Numpy counterpart. I don't see why standardization should not work for tabular data. Thus we could improve the documentation I assume. What do you think?

The `broadcastable_mean_std` can handle the tabular case already.

File: art/preprocessing/standardisation_mean_std/standardisation_mean_std.py (hunk 10 above)
Comment (hesseltuinhof) [reply art/preprocessing/standardisation_mean_std/standardisation_mean_std.py:73]: See above.

File: art/preprocessing/standardisation_mean_std/standardisation_mean_std.py (hunk 11 above)
Comment (hesseltuinhof) [reply art/preprocessing/standardisation_mean_std/standardisation_mean_std.py:90]: See above.

File: art/preprocessing/standardisation_mean_std/standardisation_mean_std_tensorflow.py (hunk 5 above)
Comment (hesseltuinhof) [reply art/preprocessing/standardisation_mean_std/standardisation_mean_std_tensorflow.py]: As discussed in our chat, standardization works for different input shapes as well. Therefore I have updated docstrings accordingly.

File: art/preprocessing/standardisation_mean_std/standardisation_mean_std_tensorflow.py (hunk 4 above)
Comment (hesseltuinhof) [reply art/preprocessing/standardisation_mean_std/standardisation_mean_std_tensorflow.py]: As discussed in our chat, I have added lazy attributes for broadcastable `mean` and `std`.

**Comment by hesseltuinhof** [comment]
@beat-buesser Thanks for your feedback. I have added your proposed change. Especially the addition of lazy attributes for broadcastable mean and standard deviation. PR has been rebased accordingly :rocket:

**Review by beat-buesser — APPROVED** [review]

**Comment by beat-buesser** [comment]
Hi @hesseltuinhof Thank you very much! This will make a long standing issue easier to solve!

**Pull request merged**